KAFKA-14214: Convert StandardAuthorizer to copy-on-write

cmccabe opened the pull request:
Convert StandardAuthorizer to use a copy-on-write sorted array rather than a ConcurrentSkipList. The issue with the skiplist was that because it was modified while in use by StandardAuthorizer#authorize, we could sometimes expose an inconsistent state. For example, if we added a "deny principal foo", followed by "allow all", a request for principal foo might happen to see the second one, without seeing the first one, even though the first one was added first.

The sorted array data structure is more compact than the tree, although it has the disadvantage of needing to be copied every time it's modified. On the broker side, we can apply a batch of updates atomically now. On the controller side, we still apply updates one-by-one, for now.

This PR renames AclAuthorizerBenchmark to AuthorizerBenchmark and extends it to report information about StandardAuthorizer as well as AclAuthorizer.

Co-authored-by: Akhilesh Chaganti <achaganti@confluent.io>

mumrah commented on metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizerData.java:
```diff
@@ -397,18 +387,34 @@ private MatchingAclRule findAclRule(
+                new StandardAclWithId(Uuid.ZERO_UUID, exemplar),
+                StandardAclWithId.ACL_COMPARATOR);
+        // Arrays.binarySearch returns a positive number if it found an exact match, and
+        // a negative number otherwise.
```
Might comment about what the negative return value indicates. It helps L404 make more sense :)

mumrah commented on metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizerData.java:
```diff
@@ -182,59 +173,58 @@ StandardAuthorizerData copyWithNewConfig(int nodeId,
+        Set<Uuid> removedAclIds
+    ) {
+        StandardAclWithId[] newAcls = new StandardAclWithId[
+                existingAcls.length + newAclEntries.size() - removedAclIds.size()];
```
nit: pull the size into a local so we don't break the line on the array brackets

cmccabe replied:
good idea

mumrah commented on metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizer.java:
```diff
@@ -96,18 +97,16 @@ public void completeInitialLoad(Exception e) {
-    public synchronized void loadSnapshot(Map<Uuid, StandardAcl> acls) {
-        data = data.copyWithNewAcls(acls.entrySet());
+    public synchronized void applyAclChanges(
+        Collection<Map.Entry<Uuid, StandardAcl>> newAcls,
```
Could we just take a `Collection<StandardAcl>` since the acl has the ID as a property?

akhileshchg replied:
I don't think `StandardAcl` has id with it. We have a different data structure for it`StandardAclWithId`

cmccabe replied:
StandardAcl doesn't contain an ID

cmccabe replied:
However, we could use StandardAclWithId here. That would require another copy in the case where we were loading a snapshot, so I'm not sure if it's worth it.

mumrah replied:
Oh, i must have been looking at StandardAclWithId :), this looks good as-is

mumrah commented on metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizerData.java:
```diff
@@ -109,14 +110,9 @@ public class StandardAuthorizerData {
-     * Contains all of the current ACLs sorted by (resource type, resource name).
+     * An immutable array of all the current ACLs sorted by (resource type, resource name).
      */
-    private final ConcurrentSkipListSet<StandardAcl> aclsByResource;
```
I see we are replacing the skip-list set with a sorted array. Don't we need to guard against duplicates in the array?

If we used a TreeSet here, it would be closer to the current implementation and I think it should have linear time when copying from another TreeSet

akhileshchg replied:
We're storing `Id` with the `StandardAcl`. Shouldn't that make it unique? I think since it is sorted, we can maybe have a conservative check to make sure there are no duplicate ids.

cmccabe replied:
TreeSets are quite slow because they have poor memory locality. They also use a lot more memory. Ideally we could use something like a BTree, but Java doesn't have those, unfortunately...

ijuma replied:
The downside of the current implementation is that updates can be pretty expensive, right? Can we file a JIRA to investigate this further post 3.3.0?

mumrah commented on metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizerData.java:
```diff
@@ -109,14 +110,9 @@ public class StandardAuthorizerData {
-    /**
-     * Contains all of the current ACLs indexed by UUID.
-     */
-    private final ConcurrentHashMap<Uuid, StandardAcl> aclsById;
```
Guess we don't need this anymore if we are removing `addAcl` and `removeAcl`?

mumrah reviewed (commented):
Thanks for the patch @cmccabe and @akhileshchg! Left a few initial comments

akhileshchg commented on metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizer.java:
```diff
@@ -138,6 +139,15 @@ public List<AuthorizationResult> authorize(
         return results;
     }
```
In `authorize` function we still do `StandardAuthorizerData curData = data;`. I don't think we need to do this anymore.

cmccabe replied:
The purpose of doing this is to avoid loading the volatile multiple times. Each time we load a volatile, it is expensive because it requires an interlocked instruction.

ijuma replied:
It also ensures a consistent snapshot is used for the duration of the method, right? Otherwise, one could get an updated version of the field for some of the authorize calls.

cmccabe commented:
Jenkins went down again. :(

```
java.nio.file.FileSystemException: /home/jenkins/workspace/Kafka_kafka-pr_PR-12636: No space left on device
```

akhileshchg commented on metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizerData.java:
```diff
@@ -182,59 +174,58 @@ StandardAuthorizerData copyWithNewConfig(int nodeId,
+        }
+        if (!newAclEntries.isEmpty()) {
+            int i = 0;
+            for (Entry<Uuid, StandardAcl> entry : newAclEntries) {
+                newAcls[existingAcls.length + i] = new StandardAclWithId(entry.getKey(), entry.getValue());
+                i++;
             }
```
Compared to the previous code, we're missing the check for duplicate ids.

cmccabe replied:
Duplicate IDs should not happen unless there is a bug. I do wish we could check for it here, but it would be very inefficient to do so, since we'd have to scan the whole array.

akhileshchg reviewed (commented):
Partial review.

mumrah commented on metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizerData.java:
```diff
@@ -182,59 +174,58 @@ StandardAuthorizerData copyWithNewConfig(int nodeId,
 
-    void addAcl(Uuid id, StandardAcl acl) {
-        try {
-            StandardAcl prevAcl = aclsById.putIfAbsent(id, acl);
-            if (prevAcl != null) {
-                throw new RuntimeException("An ACL with ID " + id + " already exists.");
-            }
-            if (!aclsByResource.add(acl)) {
-                aclsById.remove(id);
-                throw new RuntimeException("Unable to add the ACL with ID " + id +
-                    " to aclsByResource");
-            }
```
I think we lost this existing ID and duplicate ACL check in the new array code.

cmccabe replied:
Like I said above, we'd have to scan the whole array since we don't have a map from id -> acl any more. I don't think it's worth it just for a sanity check

mumrah replied:
Got it. Looking at AclControlManager, we should not generate a record for an existing ACL (as determined by the hash set in AclControlManager). 

The replace-by-ID case via `ClusterMetadataAuthorizer#addAcl` confused me since I don't see how a caller could pass in the same ID for a new ACL. The doc used to say "Add a new ACL. Any ACL with the same ID will be replaced." Anyways, it's gone now with this PR.

mumrah commented on metadata/src/main/java/org/apache/kafka/metadata/authorizer/ClusterMetadataAuthorizer.java:
```diff
@@ -76,14 +78,15 @@ public interface ClusterMetadataAuthorizer extends Authorizer {
+     * @param removedAclIds     The ACL IDs to remove.
      */
-    void removeAcl(Uuid id);
+    void applyAclChanges(
```
We should document that this method does not expect duplicates or allow replacing ACL by ID.

mumrah commented on metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizer.java:
```diff
@@ -138,6 +139,15 @@ public List<AuthorizationResult> authorize(
     }
 
+    @Override
+    public AuthorizationResult authorizeByResourceType(
```
If I understand correctly, this implementation was added to take advantage of the new binary search approach in the ACL array. IOW, an optimization over the default `authorizeByResourceType` impl?

cmccabe replied:
mainly I added this because the benchmarks looked really bad without it 😬

mumrah commented on metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizerData.java:
```diff
@@ -182,59 +174,58 @@ StandardAuthorizerData copyWithNewConfig(int nodeId,
+        if (!newAclEntries.isEmpty()) {
+            int i = 0;
+            for (Entry<Uuid, StandardAcl> entry : newAclEntries) {
+                newAcls[existingAcls.length + i] = new StandardAclWithId(entry.getKey(), entry.getValue());
```
Should this index be offset by the number we removed?

mumrah reviewed (approved):
Thanks for the updates, @cmccabe. The List implementation in StandardAuthorizerData LGTM, as does this rest of this PR. I imagine our existing test suite should give pretty good coverage of these code paths, so I'll say LGTM pending tests :) 

Thanks again to @akhileshchg for finding this issue -- it was a subtle one!

ijuma commented:
Thanks for the PR. @cmccabe can you please include the benchmark results before/after this change?

cmccabe commented:
I uploaded a new version at https://github.com/apache/kafka/pull/12662

cmccabe committed:
KAFKA-14214: Convert StandardAuthorizer to copy-on-write

Convert StandardAuthorizer to use a copy-on-write sorted array rather than a ConcurrentSkipList.
The issue with the skiplist was that because it was modified while in use by
StandardAuthorizer#authorize, we could sometimes expose an inconsistent state. For example, if we
added a "deny principal foo", followed by "allow all", a request for principal foo might happen to
see the second one, without seeing the first one, even though the first one was added first.

The sorted array data structure is more compact than the tree, although it has the disadvantage of
needing to be copied every time it's modified. On the broker side, we can apply a batch of updates
atomically now. On the controller side, we still apply updates one-by-one, for now.

This PR renames AclAuthorizerBenchmark to AuthorizerBenchmark and extends it to report information
about StandardAuthorizer as well as AclAuthorizer.

Co-authored-by: Akhilesh Chaganti achaganti@confluent.io
